### PR TITLE
Disable merging a WIP Pull request

### LINKED ISSUE #4453: Allow PRs to be marked as WIP (thus preventing them to be accidentally merged)
One thing I really miss from GitLab (I think this feature is non-existent on GitHub) is the ability to mark a pull request as WIP (Work in Progress). All GitLab MRs whose title begins with `WIP:` is blocked from being merged.

I don’t know if this would fit the plans for Gitea, but it could be definitely helpful for some.

### PR DESCRIPTION
**Pull request by JulienTant**
image

This PR fixes #4453 

The idea behind this pull request is that when the TITLE  of the pull requests starts with `[wip]` or `wip:`, then the nobody can merge that pull request until the "wip" indicator is removed from the title 

- [x] Disable the ability to merge when the PR contains the given prefixes in the UI
- [x] Disable the ability to merge when the PR contains the given prefixes in the API
- [x] Write tests
- [x] Find a way to inject settings in js
- [x] Add documentation

### CODE REVIEW

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -829,13 +829,15 @@ pulls.tab_files = Files Changed
...
 pulls.has_merged = The pull request has been merged.
+pulls.work_in_progress = "This is a work in progress. Remove the <strong>%s</strong> prefix from the title when it's ready"
 pulls.data_broken = This pull request is broken due to missing fork information.
 pulls.is_checking = "Merge conflict checking is in progress. Try again in few moments."
 pulls.can_auto_merge_desc = This pull request can be merged automatically.
 pulls.cannot_auto_merge_desc = This pull request cannot be merged automatically due to conflicts.
 pulls.cannot_auto_merge_helper = Merge manually to resolve the conflicts.
 pulls.no_merge_desc = This pull request cannot be merged because all repository merge options are disabled.
 pulls.no_merge_helper = Enable merge options in the repository settings or merge the pull request manually.
+pulls.no_merge_wip = "This pull request cannot be merged because it is marked as being a work in progress."
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]: shoud be `can not`

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -829,13 +829,15 @@ pulls.tab_files = Files Changed
 pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
 pulls.merged = Merged
 pulls.has_merged = The pull request has been merged.
+pulls.work_in_progress = "This is a work in progress. Remove the <strong>%s</strong> prefix from the title when it's ready"
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]: `This is a work in progress.` would better be `This Pull Request is marked as a work in progress.`

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -829,13 +829,15 @@ pulls.tab_files = Files Changed
 pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
 pulls.merged = Merged
 pulls.has_merged = The pull request has been merged.
+pulls.work_in_progress = "This Pull Request is marked as a work in progress. Remove the <strong>%s</strong> prefix from the title when it's ready"
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]: sorry my mistake. `Pull Request` should be `pull request` like in other messages

**Comment by JulienTant** [comment]
I added some tests + a helper under the PR title input to add `WIP:` in one click :-)

**Comment by JulienTant** [comment]
Do you guys can help me about the documentation ? I can't decide where to put a paragraph on that feature. Should I document it  ?

**Comment by JulienTant** [comment]
# Dealing with "Work In Progress" pull requests

Marking a pull request as being a work in progress will prevent that pull request from being accidentally merged. To mark a pull request as being a work in progress, you must prefix its title by `WIP:` or `[WIP]` (case insensitive)

File: models/pull.go [hunk 4]
```diff
@@ -1195,6 +1195,43 @@ func (pr *PullRequest) checkAndUpdateStatus() {
 	}
 }
 
+// Enumerate the prefixes that will help determine if a pull request is a Work In Progress
+var pullRequestWorkInProgressPrefixes = [...]string{
```
Comment (lunny) [inline models/pull.go]: I thinik it's better to set these as default and put them to app.ini files.

**Comment by lunny** [comment]
@JulienTant I think you can add a new page under `usage` named `pull request` and put a section there.

File: docs/content/doc/usage/pull-request.en-us.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    name: "Pull Request"
+    weight: 13
+    identifier: "pull-request"
+---
+
+# Pull Request
+
+## "Work In Progress" pull requests
+
+Marking a pull request as being a work in progress will prevent that pull request from being accidentally merged. To mark a pull request as being a work in progress, you must prefix its title by `WIP:` or `[WIP]` (case insensitive). Those values are configurable in your `app.ini` file :
```
[the remark this answers is not in the document]
Comment (JulienTant) [reply docs/content/doc/usage/pull-request.en-us.md:20]: i'm pretty sure that `its` is correct, http://grammarist.com/spelling/its-its/

[the remark this answers is not in the document]
Comment (adelowo) [reply docs/content/doc/usage/pull-request.en-us.md:20]: Oops... Thanks for the link

File: public/js/index.js [hunk 6]
```diff
@@ -1655,6 +1655,27 @@ function u2fRegisterRequest() {
...
 }
 
+function initWipTitle() {
+    $(".title_wip_desc > a").click(function (e) {
+        e.preventDefault();
+
+        var $issueTitle = $("#issue_title");
+        var value = $issueTitle.val().trim().toUpperCase();
+
+        var addPrefix = true;
```
Comment (lafriks) [inline public/js/index.js]: There is no need for this variable. Can be changed so:
```js
        for (var i in wipPrefixes) {
            if (value.startsWith(wipPrefixes[i].toUpperCase())) {
                return;
            }
        }

        $issueTitle.val(wipPrefixes[0] + " " + $issueTitle.val());
```

File: integrations/api_pull_test.go [hunk 7]
```diff
@@ -28,3 +30,20 @@ func TestAPIViewPulls(t *testing.T) {
 	expectedLen := models.GetCount(t, &models.Issue{RepoID: repo.ID}, models.Cond("is_pull = ?", true))
 	assert.Len(t, pulls, expectedLen)
 }
+
+// TestAPIMergePullWIP ensures that we can't merge a WIP pull request
+func TestAPIMergePullWIP(t *testing.T) {
+	prepareTestEnv(t)
+	repo := models.AssertExistsAndLoadBean(t, &models.Repository{ID: 1}).(*models.Repository)
+	owner := models.AssertExistsAndLoadBean(t, &models.User{ID: repo.OwnerID}).(*models.User)
+	pr := models.AssertExistsAndLoadBean(t, &models.PullRequest{Status: models.PullRequestStatusMergeable, HasMerged: false}).(*models.PullRequest)
```
Comment (lafriks) [inline integrations/api_pull_test.go]: Does this PR really have WIP prefix?

**Comment by lunny** [comment]
LGTM

File: options/locale/locale_en-US.ini [hunk 8]
```diff
@@ -841,13 +841,16 @@ pulls.tab_files = Files Changed
 pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
 pulls.merged = Merged
 pulls.has_merged = The pull request has been merged.
+pulls.title_wip_desc = '<a href="#">Start the title with <strong>%s</strong></a> to prevent the pull request from being merged accidentally.'
+pulls.cannot_merge_work_in_progress = "This pull request is marked as a work in progress. Remove the <strong>%s</strong> prefix from the title when it's ready"
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]: No need for quoting with `"`

**Review by lafriks — COMMENTED** [review]
Except for translation file quoting otherwise LG-TM

**Comment by JulienTant** [comment]
I decided to try everything one last time & found a bug, that I fixed: the link bellow the title bar (see bellow) was not working because of a refactoring I made on names :/ This is now fix!
image

I also remove the useless quote as requested by @lafriks

**Review by lafriks — APPROVED** [review]
Thanks, great work!

**Pull request merged**